modules/javafx.graphics/src/main/java/com/sun/glass/ui/GlassRobot.java

```diff
@@ -0,0 +1,363 @@
+     */
+    public abstract void keyRelease(KeyCode keyCode);
+
+    /**
+     * Returns the current mouse x-position.
+     *
+     * @return the current mouse x-position
+     */
+    public abstract double getMouseX();
+
+    /**
+     * Returns the current mouse y-position.
+     *
+     * @return the current mouse y-position
+     */
+    public abstract double getMouseY();
+
+    /**
+     * Moves the mouse to the specified (x,y) screen coordinates.
+     *
+     * @param x screen coordinate x to move the mouse to
+     * @param y screen coordinate y to move the mouse to
+     */
+    public abstract void mouseMove(double x, double y);
+
+    /**
+     * Presses the specified {@link MouseButton}s.
+     *
+     * @param buttons the mouse buttons to press
+     */
+    public abstract void mousePress(MouseButton... buttons);
+
+    /**
+     * Releases the specified {@link MouseButton}s.
+     *
+     * @param buttons the mouse buttons to release
+     */
+    public abstract void mouseRelease(MouseButton... buttons);
+
+    /**
+     * Scrolls the mouse wheel by the specified amount. Positive {@code wheelAmt}s
+     * scroll up whereas negative scroll down.
+     *
+     * @param wheelAmt the (signed) amount to scroll the wheel
+     */
+    public abstract void mouseWheel(int wheelAmt);
+
+    /**
+     * Returns the {@link Color} of the pixel at the specified screen coordinates
+     * of the primary screen.
+     *
+     * @param x the x coordinate to get the pixel color from
+     * @param y the y coordinate to get the pixel color from
+     * @return the pixel color at the specified screen coordinates
+     */
+    public abstract Color getPixelColor(double x, double y);
+
+    /**
+     * Captures the specified rectangular area of the screen and uses it to fill the given
+     * {@code data} array with the raw pixel data. The data is in RGBA format where each
+     * pixel in the image is encoded as 4 bytes - one for each color component of each
+     * pixel. If this method is not overridden by subclasses then
+     * {@link #getScreenCapture(WritableImage, double, double, double, double, boolean)}
+     * must be overridden to not call this method.
+     *
+     * @param x the starting x-position of the rectangular area to capture
+     * @param y the starting y-position of the rectangular area to capture
+     * @param width the width of the rectangular area to capture
+     * @param height the height of the rectangular area to capture
+     * @param data the array to fill with the raw pixel data corresponding to
+     * the captured region
+     * @param scaleToFit If {@literal true} the returned {@code Image} will be
+     * scaled to fit the request dimensions, if necessary. Otherwise the size
+     * of the returned image will depend on the output scale (DPI) of the primary
+     * screen.
+     */
+    public void getScreenCapture(int x, int y, int width, int height, int[] data, boolean scaleToFit) {
```

Comment:
Why the output will depend from the scale of the primary screen? I guess it should depend from the screen where this rectangle is located, no? Same for the method below.

Reply:
Why this method is not abstract but throws an exception? If not all peer will implement it then this exception should be specified.

Reply:
Does JavaFX actually support having multiple monitors each with different DPIs? If that's the case then the wording should definitely be changed, although I am unsure how the implementations work in that regard.

Reply:
Yes, JavaFX does support multiple monitors with different DPI.

Regarding the exception, this an implementation class, so doesn't need to be fully specified. Maybe changing it to throw InternalError rather than an exception would help clarify that if this is ever thrown, then the Glass platform is coded incorrectly (at least one of the two methods must be overridden).

Reply:
I think that the logic which scales the region should be similar to the logic which changes the bounds of the window, but I do not know maybe setbounds does not work as well =(

Reply:
That might make sense as a follow-up, especially if it turns out that screen capture has a problem with multiple screens that have different DPI scales.